Add io.github.Adda0.Stardrop

Adda0 opened the pull request:
This PR adds [flatpak support](https://github.com/Adda0/stardrop-flatpak) for an unofficial [fork of Stardrop](https://github.com/Adda0/Stardrop/) Mod Manager for Stardew Valley with custom changes, with the [permission of the upstream Stardrop author](https://github.com/Floogen/Stardrop/issues/79#issuecomment-1862709718), adhering to the GPLv3 licence requirements.

### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I have [built][build] and tested the submission locally.
    - Ignoring the screenshot error as instructed for new flatpaks. But I cannot figure out where to set the developer name so that appstream can find it. Will fix it if I find how to do that.
    ```
    {
    "errors": [
        "appstream-screenshots-not-mirrored-in-ostree"
    ],
    "warnings": [
        "appstream-missing-developer-name"
    ]
    }
    ``` 
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
    - Stardrop allows for storing added mods in any user-specified folder. Therefore, the flatpak has host filesystem access. I am thinking about how to restrict the access further, but that is still a work in progress (hence the commented out different filesystem args). 
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** [permission of the upstream Stardrop author](https://github.com/Floogen/Stardrop/issues/79#issuecomment-1862709718), adhering to the GPLv3 licence requirements.
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*
    - Work in progress to eventually merge the custom changes into the upstream repo and later move this flatpak to the upstream repo, too. 

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

bbhtt commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+app-id: io.github.Adda0.Stardrop
+runtime: org.freedesktop.Platform
+runtime-version: '22.08'
```
```suggestion
runtime-version: '23.08'
```

Adda0 replied:
Fixed.

bbhtt commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+        path: io.github.Adda0.Stardrop.desktop
+      - type: file
+        path: Stardrop_512x512x32.png
+finish-args:
+  # - "--socket=wayland"
+  # - "--socket=fallback-x11"
+  - "--socket=x11"
+  - "--share=ipc"
+  - "--share=network"
+  - "--device=dri"
+  # - "--filesystem=~/.local/share/Steam/steamapps/common/Stardew\ Valley/"
+  # - "--filesystem=home"
+  - "--filesystem=host"
+  # - "--filesystem=xdg-download"
+  - "--socket=pulseaudio"
+command: "Stardrop"
```
This should be on top

Adda0 replied:
Done.

bbhtt commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+        PKG_CONFIG_PATH: /app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet7/lib/pkgconfig
+modules:
+  # Taken from https://github.com/flathub/com.unity.UnityHub/blob/af171c1ee22994e7ef698004bd44ba7e8bc71cf0/openssl-1.1.yaml.
+  - name: openssl-1.1
```
openssl 1.1 is no longer developed.

Adda0 replied:
Unfortunately, it is. It is a long-standing [issue](https://github.com/Floogen/Stardrop/issues?q=is%3Aissue+libssl) with Stardrop. I would like to update all the dependencies in order to not need libssl1.1 any more, too, but that is still WIP.

bbhtt commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+    build-commands:
+      - "install-sdk.sh"
+      - "dotnet publish -c Release --source ./nuget-sources Stardrop/Stardrop.csproj --runtime linux-x64 --self-contained true"
+      - cp -r /run/build/Stardrop/Stardrop/bin/Release/linux-x64/publish/* /app/bin/
```
doesn't this make it x86_64 only?

Adda0 replied:
I modified the command to hopefully include all architectures where the flatpak might be run.

bbhtt commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+      - install -Dm644 io.github.Adda0.Stardrop.desktop /app/share/applications/io.github.Adda0.Stardrop.desktop
+      - install -Dm644 Stardrop_512x512x32.png /app/share/icons/hicolor/512x512/apps/io.github.Adda0.Stardrop.png
+    sources:
+      - type: file
+        path: io.github.Adda0.Stardrop.metainfo.xml
+      - type: file
+        path: io.github.Adda0.Stardrop.desktop
+      - type: file
+        path: Stardrop_512x512x32.png
```
these do not exist here? Also, desktop file, icon and appdata should come from the upstream, you should submit them.

Adda0 replied:
Just to clarify, by submitting, you just mean pushing all of them together with `nuget-sources.json` to this PR and the paths here in the manifest file remain the same, i.e., in the current directory. Is that right?

Adda0 replied:
Files are submitted in this PR.

bbhtt commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+      - "dotnet publish -c Release --source ./nuget-sources Stardrop/Stardrop.csproj --runtime linux-x64 --self-contained true"
+      - cp -r /run/build/Stardrop/Stardrop/bin/Release/linux-x64/publish/* /app/bin/
+    sources:
+      - nuget-sources.json
```
this does not exist?

Adda0 replied:
It is referring to this list of sources: [nuget-sources.json](https://github.com/Adda0/stardrop-flatpak/blob/main/nuget-sources.json). Should I include this list of sources in this PR already?

bbhtt replied:
> Should I include this list of sources in this PR already?

yes

hfiguiere replied:
how do you think it finds the file?

Adda0 replied:
File is submitted in this PR.

bbhtt commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+  - "--device=dri"
+  # - "--filesystem=~/.local/share/Steam/steamapps/common/Stardew\ Valley/"
+  # - "--filesystem=home"
+  - "--filesystem=host"
```
host is too broad

> Stardrop allows for storing added mods in any user-specified folder.

What are these folders?

hfiguiere replied:
if it's just to install mods, `host` easily too broad.

Adda0 replied:
That is the problem. Stardrop allows the users to set their own directory where they store their mods. From what I have seen, some put the mods "installed" by Stardrop (extracting user-downloaded archive of the mod) in the game main mods folder (that is the only specific folder I can directly request access to), others put them somewhere in their home directory, others possibly who knows where.

I could possibly say that access to home directory is enough and hope that everyone will have their mods stored under their home directory. However, I think this should be OK even for Steam Decks where the Steam directory is under the home directory somewhere.

Would `home` be strict enough? Or is there a better way to let the user choose an arbitrary folder where Stardrop will then have read-write access with the ability to change the directory afterwards?

bbhtt commented:
> But I cannot figure out where to set the developer name so that appstream can find it. Will fix it if I find how to do that.

In the appdata file add `<developer_name>foo</developer_name>`

hfiguiere commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+    append-path: /usr/lib/sdk/dotnet7/bin
+    append-ld-library-path: /usr/lib/sdk/dotnet7/lib
+    env:
+        PKG_CONFIG_PATH: /app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet7/lib/pkgconfig
```
please use `prepend-pkg-config-path`.

hfiguiere replied:
yes the doc for the SDK needs to be updated.

Adda0 replied:
Fixed.

hfiguiere commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+      - type: archive
+        url: "https://github.com/Adda0/Stardrop/archive/refs/tags/v1.0.0-beta.38-custom_adda.1.tar.gz"
+        sha256: eb1a31fa533ef0d3f4f95d49bb3f7ff1b6692307fb1d0aeb7c21b0b7debcad8d
+      # - type: dir
+      #   path: ../Stardrop-repo/
```
```suggestion
```

Adda0 replied:
Done.

hfiguiere commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+      - type: file
+        path: Stardrop_512x512x32.png
+finish-args:
+  # - "--socket=wayland"
+  # - "--socket=fallback-x11"
```
```suggestion
```

Adda0 replied:
Done.

hfiguiere commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,76 @@
+        path: io.github.Adda0.Stardrop.desktop
+      - type: file
+        path: Stardrop_512x512x32.png
+finish-args:
```
finish-args is usually put above modules.

Adda0 replied:
Done.

Adda0 commented:
Thank you very much for your reviews and advices. I addressed all the issues. Please, feel free to close those you consider resolved and further comment on those that remain unresolved.

> >  But I cannot figure out where to set the developer name so that appstream can find it. Will fix it if I find how to do that.
>
> In the appdata file add <developer_name>foo</developer_name>

I now have both `<developer><name>foo</name></developer>` and `<developer_name></developer_name>` next to each other. Is that correct?

hfiguiere commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,68 @@
+    buildsystem: simple
+    build-commands:
+      - "install-sdk.sh"
+      - "dotnet publish -c Release --source ./nuget-sources Stardrop/Stardrop.csproj --runtime linux-x64 --self-contained true"
```
this still looks specific to x86_64.

Adda0 replied:
You are right. I missed this one. `dotnet` should pick the correct runtime without explicitly specifying `--runtime` argument, so I just do not specify the runtime now.

bbhtt commented on io.github.Adda0.Stardrop.metainfo.xml:
```diff
@@ -0,0 +1,52 @@
+  <launchable type="desktop-id">io.github.Adda0.Stardrop.desktop</launchable>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://github.com/Adda0/stardrop-flatpak/blob/main/screenshots/main-screenshot-resized.png?raw=true</image>
+    </screenshot>
+    <screenshot>
+      <image>https://github.com/Adda0/stardrop-flatpak/blob/main/screenshots/updates-settings-screenshot-resized.png?raw=true</image>
+    </screenshot>
+    <screenshot>
+      <image>https://github.com/Adda0/stardrop-flatpak/blob/main/screenshots/search-author-mods-screenshot-resized.png?raw=true</image>
```
Screenshots should be from a tag/commit

Adda0 replied:
Fixed. Linked them from a specific commit.

bbhtt commented on io.github.Adda0.Stardrop.yaml:
```diff
@@ -0,0 +1,68 @@
+build-options:
+    append-path: /usr/lib/sdk/dotnet7/bin
+    append-ld-library-path: /usr/lib/sdk/dotnet7/lib
+    prepend-pkg-config-path: /app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet7/lib/pkgconfig
```
```suggestion
    prepend-pkg-config-path: /usr/lib/sdk/dotnet7/lib/pkgconfig
```

bbhtt commented:
bot, build io.github.Adda0.Stardrop

bbhtt commented on io.github.Adda0.Stardrop.metainfo.xml:
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+  <id>io.github.adda0.stardrop</id>
```
Is it `adda0` or `Adda0`?

hfiguiere replied:
It's case sensitive here and has to match the filename.

```suggestion
  <id>io.github.Adda0.Stardrop</id>
```

Adda0 replied:
The suggested fix was applied.

bbhtt commented:
aarch64 build is failing. Looks like you need to add aarc64 sources in nuget-sources.json

```
MSBuild version 17.7.4+3ebbd7c49 for .NET
  Determining projects to restore...
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Runtime.linux-arm64 with version (= 7.0.14)
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1102:   - Found 1 version(s) in /run/build/Stardrop/nuget-sources [ Nearest version: 8.0.0 ]
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1102: Unable to find package Microsoft.AspNetCore.App.Runtime.linux-arm64 with version (= 7.0.14)
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1102:   - Found 1 version(s) in /run/build/Stardrop/nuget-sources [ Nearest version: 8.0.0 ]
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1101: Unable to find package Microsoft.NETCore.App.Host.linux-arm64. No packages exist with this id in source(s): /run/build/Stardrop/nuget-sources
  Failed to restore /run/build/Stardrop/Stardrop/Stardrop.csproj (in 1.91 sec).
FB: host_command_exited_cb 633672 256
```

Adda0 commented:
The dependencies were added. Can I build flatpak for all architectures locally somehow to test the build for arm64? `flatpak run org.flatpak.Builder` has argument `--arch` which says that the architecture building for must be host compatible, which is not the case on my machine.

hfiguiere commented:
bot, build io.github.Adda0.Stardrop

Adda0 commented:
I tried to modify Stardrop project to build for arm64, too. However, I don't know how to test the build for flatpak for arm64 locally. Stardrop builds for arm64 correctly, it seems.

barthalion commented:
bot, build io.github.Adda0.Stardrop

barthalion commented:
/merge
